ACP-196: Precompile for secp256r1 Curve Support (#204)
avalanche-foundation/ACPs: ACPs/196-precompile-secp256r1/README.md

# ACP-196: Precompile for secp256r1 Curve Support

| ACP | 196 |
| :--- | :--- |
| **Title** | Precompile for secp256r1 Curve Support |
| **Author(s)** | [Santiago Cammi](https://github.com/scammi), [Arran Schlosberg](https://github.com/ARR4N)  |
| **Status** | Proposed ([Discussion](https://github.com/avalanche-foundation/ACPs/discussions/XXX)) |
| **Track** | Standards |

## Abstract

> This proposal introduces a precompiled contract that performs signature verifications for the secp256r1 elliptic curve on Avalanche's C-Chain and Subnet-EVM chains. The precompile will be implemented at address `0x0000000000000000000000000000000000000100` and will enable native verification of P-256 signatures, significantly improving gas efficiency for biometric authentication systems, WebAuthn, and modern device-based signing mechanisms.

Reviewer: ```suggestion
This proposal introduces a precompiled contract that performs signature verifications for the secp256r1 elliptic curve on Avalanche's C-Chain. The precompile will be implemented at address `0x0200000000000000000000000000000000000006` and will enable native verification of P-256 signatures, significantly improving gas efficiency for biometric authentication systems, WebAuthn, and modern device-based signing mechanisms.
```

L1s (previously called subnets) are sovereign and can add their own precompiles as they wish, so this is just for C-Chain. We use different address ranges and I _think_ I've set the next one correctly, but we can check that later.

## Motivation

> The secp256r1 (P-256) elliptic curve is the standard cryptographic curve used by modern device security systems, including Apple's Secure Enclave, Android Keystore, WebAuthn, and Passkeys. However, Avalanche currently only supports secp256k1 natively, forcing developers to use expensive Solidity-based verification that costs 200k-330k gas per signature verification.

Reviewer: Where is the 200k-330k estimate from?

Author: There are a few smart contract based verifiers implemented, the following posts goes through each of the solutions and display how much each cost.

https://hackmd.io/@1ofB8klpQky-YoR5pmPXFQ/SJ0nuzD1T#Smart-Contract-Based-Verifiers

https://www.alchemy.com/blog/what-is-rip-7212#how-does-rip-7212-compare-to-other-solutions

This ACP proposes implementing EIP-7212's secp256r1 precompiled contract to unlock significant ecosystem benefits:

### **Enterprise & Institutional Adoption**
- **Reduced onboarding friction**: Enterprises can leverage existing biometric authentication infrastructure instead of managing seed phrases or hardware wallets
- **Regulatory compliance**: Institutions can utilize their approved device security standards and identity management systems
- **Cost optimization**: 100x gas reduction (from 200k-330k to 3,450 gas) makes enterprise-scale applications economically viable

The 100x gas cost reduction makes these use cases economically viable while maintaining the security properties institutions and users expect from their existing devices.

## Specification

> This ACP implements [EIP-7212](https://eips.ethereum.org/EIPS/eip-7212) for secp256r1 signature verification on Avalanche. The specification follows EIP-7212 exactly, with the precompiled contract deployed at address `0x0000000000000000000000000000000000000100`.

Reviewer: ```suggestion
This ACP implements [EIP-7212](https://eips.ethereum.org/EIPS/eip-7212) for secp256r1 signature verification on Avalanche. The specification follows EIP-7212 exactly, with the precompiled contract deployed at address `0x0200000000000000000000000000000000000006`.
```

> This ACP implements [EIP-7212](https://eips.ethereum.org/EIPS/eip-7212) for secp256r1 signature verification on Avalanche. The specification follows EIP-7212 exactly, with the precompiled contract deployed at address `0x0000000000000000000000000000000000000100`.

Reviewer: I think this will need to be a relatively standalone document, not reliant on something that can change in another ecosystem. EIP-7212 has already become RIP-7212 and the contents may change too.

Author: You recommend we include the full technical specification directly in the ACP ?

Reviewer: I think so. @meaghanfitzgerald knows best though. How much needs to be included in an ACP vs referenced?

### Core Functionality
- **Input**: 160 bytes (message hash + signature components r,s + public key coordinates x,y)

> - **Output**: 32 bytes (success: `0x...01`, failure: empty)  

Reviewer: > failure: empty

Does that mean 32 zero bytes, or no bytes at all?

> - **Output**: 32 bytes (success: `0x...01`, failure: empty)  

Reviewer: > **Output**: 32 bytes

This will be in the `returndata` buffer, which would typically be copied to memory and not the stack, so I don't think it's necessary to use an entire word as this increases the memory usage of the caller.

Author: Yes, I agree, but it breaks compatibility with other chains' EIP-7212 implementations.

Is this something we care about?

I understand that this should be a standalone document, so optimizing for efficiency makes sense.

Reviewer: Good point. I'm in favour of ecosystem compatibility.

Do you know how many chains have implemented RIP-7212? With EIPs we always wait until they're live on Ethereum mainnet because then we know that it's a stable interface. I'd hate to release something, the rest of the ecosystem does something different, and we don't have shared tooling.

That said, we can always create a later ACP to add alternative interfaces if necessary. Same precompile, multiple ways to interact with it.

- **Gas Cost**: 3,450 gas (based on EIP-7212 benchmarking)
- **Validation**: Full compliance with NIST FIPS 186-5 specification

### Activation
This precompile may be activated as part of Avalanche's next network upgrade. Individual Avalanche L1s and subnets could adopt this enhancement independently through their respective client software updates.

For complete technical specifications, validation requirements, and implementation details, refer to [EIP-7212](https://eips.ethereum.org/EIPS/eip-7212).

## Backwards Compatibility

This ACP introduces a new precompiled contract and does not modify existing functionality. No backwards compatibility issues are expected since:

> 1. The precompile uses a previously unused address (`0x100`)

Reviewer: ```suggestion
1. The precompile uses a previously unused address (`0x0200000000000000000000000000000000000006`)
```

2. No existing opcodes or consensus rules are modified
3. The change is additive and opt-in for applications

Adoption requires a coordinated network upgrade for the C-Chain. Subnet-EVM chains can adopt this enhancement independently by upgrading their client software.

## Security Considerations

### Cryptographic Security
- The secp256r1 curve is standardized by NIST and widely vetted
- Security properties are comparable to secp256k1 (used by ECRECOVER)

> - Implementation follows NIST FIPS 186-5 specification exactly

Reviewer: > - Implementation follows NIST FIPS 186-5 specification exactly

Which implementation does? [This](https://pkg.go.dev/crypto/elliptic#P256) one?

Author: To be accurate, Go's implementation actually follows FIPS 186-3. But the P-256 curve itself is identical across FIPS 186-3, 186-4, and 186-5. The core mathematical parameters, signature verification algorithms, and implementation approach are the same.

I'm specifying we are using the version FIPS 186-3, as that's what the Go implementation uses.

### Implementation Security  
- Signature verification (not recovery) approach maximizes compatibility with existing P-256 ecosystem

> - No malleability check included to match NIST specification, but wrapper libraries should add this

Reviewer: Why? If every wrapper library adds it then every single one needs to be correct, instead of the precompile just being correct.

@yacovm please can you give an opinion?

Reviewer 2: > No malleability check included to match NIST specification

I think it's a bit non uniform that we check for malleability in the precompile with one curve (secp256k1) but not with the other. 

It would be interesting to look whether other networks have this deviation or not.

Reviewer: SGTM 👍

Reviewer 3: Ah, the comment was about the ecrecover precompile. I thought we were referring to transaction signatures when discussing secp256k1.

sgtm.

- Input validation prevents invalid curve points and out-of-range signature components

### Network Security
- Gas cost prevents potential DoS attacks through expensive computation
- Precompile implementation should use battle-tested cryptographic libraries
- No consensus-level security implications beyond standard precompile considerations

## Reference Implementation

The implementation will build upon existing work:

> 1. **EIP-7212 Reference**: The [go-ethereum implementation](https://github.com/ethereum/go-ethereum) of EIP-7212 provides the foundation

Reviewer: I can't find an implementation anywhere in their repo.

Author: [Here](https://github.com/maticnetwork/bor/pull/1069) it is. Also, updating the link.
